blockprod/src/detail/mod.rs

```diff
@@ -189,27 +173,38 @@ impl BlockProduction {
         Ok(block_header)
     }
 
-    pub async fn generate_block(
-        &mut self,
+    /// The function the creates a new block.
+    /// Returns the block and a oneshot receiver that will be notified when
+    /// the internal job is finished. Generally this can be used to ensure
+    /// that the block production process has ended and that there's no
+    /// remnants in the job manager.
+    pub async fn produce_block(
+        &self,
+        reward_destination: Destination,
+        transactions_source: TransactionsSource,
+    ) -> Result<(Block, oneshot::Receiver<usize>), BlockProductionError> {
+        self.produce_block_with_custom_id(reward_destination, transactions_source, None)
+            .await
+    }
+
+    async fn produce_block_with_custom_id(
+        &self,
         _reward_destination: Destination,
         transactions_source: TransactionsSource,
-    ) -> Result<Block, BlockProductionError> {
+        custom_id: Option<Vec<u8>>,
+    ) -> Result<(Block, oneshot::Receiver<usize>), BlockProductionError> {
         let stop_flag = Arc::new(false.into());
 
         let timestamp = BlockTimestamp::from_duration_since_epoch(self.time_getter().get_time());
 
         let (_, tip_at_start) = self.pull_consensus_data(timestamp).await?;
 
         let (job_key, mut cancel_receiver) =
-            self.job_manager.add_job(tip_at_start.block_id()).await?;
+            self.job_manager.add_job(custom_id, tip_at_start.block_id()).await?;
 
         // At the end of this function, the job has to be removed
-        // Once the job key is used, we swap it with None... this is all temporary.
-        // The docs in the function self.stop_job_in_scope()
-        let job_key_destroyed = AtomicBool::new(false);
-        let _job_remove_checker = OnceDestructor::new(|| {
-            assert!(job_key_destroyed.load(std::sync::atomic::Ordering::SeqCst));
-        });
+        let (job_remover_func, end_confirm_receiver) = self.job_manager.make_job_stopper_function();
+        let _job_remover = OnceDestructor::new(move || job_remover_func(job_key));
```

Comment:
Can these two lines be put inside `add_job()`?

Also, instead of returning the block as is, to reduce the amount of things that are returned and reduce API complexity, maybe `end_confirm_receiver` should just be `block_receiver` and so the below code of:

```
        let (block, end_receiver) = self.produce_block(reward_destination, transactions).await?;
        let _finished = end_receiver.await;
        Ok(block)
```

instead becomes something like:

```
        let block_receiver = self.produce_block(reward_destination, transactions).await?;
        let block = block_receiver.await;
        Ok(block)
```

or the uglier but shorter:

```
        let block = self.produce_block(reward_destination, transactions).await?.await;
        Ok(block)
```

Reply:
We can do this but then we'll need to create a custom future. I'm not up for this, tbh. Right now it behaves correctly and that's the concern I wanted to fix.

Feel free to add a TODO for this.